// @airtable/blocks/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.__reset = __reset;
exports.session = exports.cursor = exports.unstable_fetchAsync = exports.viewport = exports.undoRedo = exports.settingsButton = exports.runInfo = exports.reload = exports.installationId = exports.globalConfig = exports.base = exports.__sdk = void 0;

var _perform_record_action = require("./perform_record_action");

var _warning = _interopRequireWildcard(require("./warning"));

var _airtable_interface = _interopRequireDefault(require("./injected/airtable_interface"));

var _sdk2 = _interopRequireDefault(require("./sdk"));

var _create_aggregators = require("./models/create_aggregators");

var _initialize_block = require("./ui/initialize_block");

/** @internal */
var __sdk;

exports.__sdk = __sdk;
var base;
exports.base = base;
var globalConfig;
exports.globalConfig = globalConfig;
var installationId;
exports.installationId = installationId;
var reload;
exports.reload = reload;
var runInfo;
exports.runInfo = runInfo;
var settingsButton;
exports.settingsButton = settingsButton;
var undoRedo;
exports.undoRedo = undoRedo;
var viewport;
exports.viewport = viewport;
var unstable_fetchAsync; // The `cursor` binding is declared on the following line solely as a signal to
// the TypeScript compiler. The exported value is actually controlled by the
// subsequent CommonJS module property descriptor.

exports.unstable_fetchAsync = unstable_fetchAsync;
var cursor;
exports.cursor = cursor;
Object.defineProperty(module.exports, 'cursor', {
  enumerable: true,

  get() {
    (0, _warning.default)('`import {cursor} from "@airtable/blocks"` is deprecated. Use `import {useCursor} from "@airtable/blocks/ui"` instead.');
    return __sdk.cursor;
  }

}); // The `session` binding is declared on the following line solely as a signal
// to the TypeScript compiler. The exported value is actually controlled by the
// subsequent CommonJS module property descriptor.

var session;
exports.session = session;
Object.defineProperty(module.exports, 'session', {
  enumerable: true,

  get() {
    (0, _warning.default)('`import {session} from "@airtable/blocks"` is deprecated. Use `import {useSession} from "@airtable/blocks/ui"` instead.');
    return __sdk.session;
  }

});
Object.defineProperty(module.exports, 'UI', {
  enumerable: true,

  get() {
    (0, _warning.default)('`import {UI} from "@airtable/blocks"` is deprecated. Use `import * as UI from "@airtable/blocks/ui/ui"` instead.');
    return require('./ui/ui');
  }

});
Object.defineProperty(module.exports, 'models', {
  enumerable: true,

  get() {
    (0, _warning.default)('`import {models} from "@airtable/blocks"` is deprecated. Use `import * as models from "@airtable/blocks/models/models"` instead.');
    return require('./models/models');
  }

});
/** @internal */

function __reset() {
  exports.__sdk = __sdk = new _sdk2.default((0, _airtable_interface.default)());
  var _sdk = __sdk;
  exports.base = base = _sdk.base;
  exports.globalConfig = globalConfig = _sdk.globalConfig;
  exports.installationId = installationId = _sdk.installationId;
  exports.reload = reload = _sdk.reload;
  exports.runInfo = runInfo = _sdk.runInfo;
  exports.settingsButton = settingsButton = _sdk.settingsButton;
  exports.undoRedo = undoRedo = _sdk.undoRedo;
  exports.viewport = viewport = _sdk.viewport;
  exports.unstable_fetchAsync = unstable_fetchAsync = _sdk.unstable_fetchAsync;
  // The following injections are necessary to allow the referenced modules
  // to be imported directly by consumer code while also avoiding cycles in
  // the module dependency graph.
  (0, _create_aggregators.__injectSdkIntoCreateAggregators)(__sdk);
  (0, _perform_record_action.__injectSdkIntoPerformRecordAction)(__sdk);
  (0, _initialize_block.__injectSdkIntoInitializeBlock)(__sdk);
  (0, _warning.__injectSdkIntoWarning)(__sdk);
}

__reset();
