// @airtable/blocks/dist/cjs/ui/ui.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  globalAlert: true,
  GlobalAlert: true,
  BaseProvider: true,
  colors: true,
  colorUtils: true,
  Synced: true,
  TablePicker: true,
  TablePickerSynced: true,
  FieldPicker: true,
  FieldPickerSynced: true,
  ViewPicker: true,
  ViewPickerSynced: true,
  InputSynced: true,
  CellRenderer: true,
  expandRecord: true,
  expandRecordList: true,
  expandRecordPickerAsync: true,
  FieldIcon: true,
  ColorPaletteSynced: true,
  ProgressBar: true,
  RecordCard: true,
  RecordCardList: true,
  SelectSynced: true,
  SelectButtonsSynced: true,
  SwitchSynced: true,
  ViewportConstraint: true,
  initializeBlock: true,
  withHooks: true,
  useLoadable: true,
  useRecordIds: true,
  useRecords: true,
  useRecordById: true,
  useRecordQueryResult: true,
  useBase: true,
  useCursor: true,
  useSession: true,
  useSettingsButton: true,
  useSynced: true,
  useWatchable: true,
  useViewport: true,
  useGlobalConfig: true,
  useViewMetadata: true,
  useRecordActionData: true,
  registerRecordActionDataCallback: true
};
Object.defineProperty(exports, "GlobalAlert", {
  enumerable: true,
  get: function get() {
    return _global_alert.default;
  }
});
Object.defineProperty(exports, "BaseProvider", {
  enumerable: true,
  get: function get() {
    return _base_provider.default;
  }
});
Object.defineProperty(exports, "colors", {
  enumerable: true,
  get: function get() {
    return _colors.default;
  }
});
Object.defineProperty(exports, "colorUtils", {
  enumerable: true,
  get: function get() {
    return _color_utils.default;
  }
});
Object.defineProperty(exports, "Synced", {
  enumerable: true,
  get: function get() {
    return _synced.default;
  }
});
Object.defineProperty(exports, "TablePicker", {
  enumerable: true,
  get: function get() {
    return _table_picker.default;
  }
});
Object.defineProperty(exports, "TablePickerSynced", {
  enumerable: true,
  get: function get() {
    return _table_picker_synced.default;
  }
});
Object.defineProperty(exports, "FieldPicker", {
  enumerable: true,
  get: function get() {
    return _field_picker.default;
  }
});
Object.defineProperty(exports, "FieldPickerSynced", {
  enumerable: true,
  get: function get() {
    return _field_picker_synced.default;
  }
});
Object.defineProperty(exports, "ViewPicker", {
  enumerable: true,
  get: function get() {
    return _view_picker.default;
  }
});
Object.defineProperty(exports, "ViewPickerSynced", {
  enumerable: true,
  get: function get() {
    return _view_picker_synced.default;
  }
});
Object.defineProperty(exports, "InputSynced", {
  enumerable: true,
  get: function get() {
    return _input_synced.default;
  }
});
Object.defineProperty(exports, "CellRenderer", {
  enumerable: true,
  get: function get() {
    return _cell_renderer.default;
  }
});
Object.defineProperty(exports, "expandRecord", {
  enumerable: true,
  get: function get() {
    return _expand_record.default;
  }
});
Object.defineProperty(exports, "expandRecordList", {
  enumerable: true,
  get: function get() {
    return _expand_record_list.default;
  }
});
Object.defineProperty(exports, "expandRecordPickerAsync", {
  enumerable: true,
  get: function get() {
    return _expand_record_picker_async.default;
  }
});
Object.defineProperty(exports, "FieldIcon", {
  enumerable: true,
  get: function get() {
    return _field_icon.default;
  }
});
Object.defineProperty(exports, "ColorPaletteSynced", {
  enumerable: true,
  get: function get() {
    return _color_palette_synced.default;
  }
});
Object.defineProperty(exports, "ProgressBar", {
  enumerable: true,
  get: function get() {
    return _progress_bar.default;
  }
});
Object.defineProperty(exports, "RecordCard", {
  enumerable: true,
  get: function get() {
    return _record_card.default;
  }
});
Object.defineProperty(exports, "RecordCardList", {
  enumerable: true,
  get: function get() {
    return _record_card_list.default;
  }
});
Object.defineProperty(exports, "SelectSynced", {
  enumerable: true,
  get: function get() {
    return _select_synced.default;
  }
});
Object.defineProperty(exports, "SelectButtonsSynced", {
  enumerable: true,
  get: function get() {
    return _select_buttons_synced.default;
  }
});
Object.defineProperty(exports, "SwitchSynced", {
  enumerable: true,
  get: function get() {
    return _switch_synced.default;
  }
});
Object.defineProperty(exports, "ViewportConstraint", {
  enumerable: true,
  get: function get() {
    return _viewport_constraint.default;
  }
});
Object.defineProperty(exports, "initializeBlock", {
  enumerable: true,
  get: function get() {
    return _initialize_block.initializeBlock;
  }
});
Object.defineProperty(exports, "withHooks", {
  enumerable: true,
  get: function get() {
    return _with_hooks.default;
  }
});
Object.defineProperty(exports, "useLoadable", {
  enumerable: true,
  get: function get() {
    return _use_loadable.default;
  }
});
Object.defineProperty(exports, "useRecordIds", {
  enumerable: true,
  get: function get() {
    return _use_records.useRecordIds;
  }
});
Object.defineProperty(exports, "useRecords", {
  enumerable: true,
  get: function get() {
    return _use_records.useRecords;
  }
});
Object.defineProperty(exports, "useRecordById", {
  enumerable: true,
  get: function get() {
    return _use_records.useRecordById;
  }
});
Object.defineProperty(exports, "useRecordQueryResult", {
  enumerable: true,
  get: function get() {
    return _use_records.useRecordQueryResult;
  }
});
Object.defineProperty(exports, "useBase", {
  enumerable: true,
  get: function get() {
    return _use_base.default;
  }
});
Object.defineProperty(exports, "useCursor", {
  enumerable: true,
  get: function get() {
    return _use_cursor.default;
  }
});
Object.defineProperty(exports, "useSession", {
  enumerable: true,
  get: function get() {
    return _use_session.default;
  }
});
Object.defineProperty(exports, "useSettingsButton", {
  enumerable: true,
  get: function get() {
    return _use_settings_button.default;
  }
});
Object.defineProperty(exports, "useSynced", {
  enumerable: true,
  get: function get() {
    return _use_synced.default;
  }
});
Object.defineProperty(exports, "useWatchable", {
  enumerable: true,
  get: function get() {
    return _use_watchable.default;
  }
});
Object.defineProperty(exports, "useViewport", {
  enumerable: true,
  get: function get() {
    return _use_viewport.default;
  }
});
Object.defineProperty(exports, "useGlobalConfig", {
  enumerable: true,
  get: function get() {
    return _use_global_config.default;
  }
});
Object.defineProperty(exports, "useViewMetadata", {
  enumerable: true,
  get: function get() {
    return _use_view_metadata.default;
  }
});
Object.defineProperty(exports, "useRecordActionData", {
  enumerable: true,
  get: function get() {
    return _use_record_action_data.default;
  }
});
Object.defineProperty(exports, "registerRecordActionDataCallback", {
  enumerable: true,
  get: function get() {
    return _perform_record_action.registerRecordActionDataCallback;
  }
});
exports.globalAlert = void 0;

var _global_alert = _interopRequireDefault(require("./global_alert"));

require("../");

var _base_provider = _interopRequireDefault(require("./base_provider"));

var _colors = _interopRequireDefault(require("../colors"));

var _color_utils = _interopRequireDefault(require("../color_utils"));

var _synced = _interopRequireDefault(require("./synced"));

var _table_picker = _interopRequireDefault(require("./table_picker"));

var _table_picker_synced = _interopRequireDefault(require("./table_picker_synced"));

var _field_picker = _interopRequireDefault(require("./field_picker"));

var _field_picker_synced = _interopRequireDefault(require("./field_picker_synced"));

var _view_picker = _interopRequireDefault(require("./view_picker"));

var _view_picker_synced = _interopRequireDefault(require("./view_picker_synced"));

var _input_synced = _interopRequireDefault(require("./input_synced"));

var _cell_renderer = _interopRequireDefault(require("./cell_renderer"));

var _expand_record = _interopRequireDefault(require("./expand_record"));

var _expand_record_list = _interopRequireDefault(require("./expand_record_list"));

var _expand_record_picker_async = _interopRequireDefault(require("./expand_record_picker_async"));

var _field_icon = _interopRequireDefault(require("./field_icon"));

var _color_palette_synced = _interopRequireDefault(require("./color_palette_synced"));

var _progress_bar = _interopRequireDefault(require("./progress_bar"));

var _record_card = _interopRequireDefault(require("./record_card"));

var _record_card_list = _interopRequireDefault(require("./record_card_list"));

var _select_synced = _interopRequireDefault(require("./select_synced"));

var _select_buttons_synced = _interopRequireDefault(require("./select_buttons_synced"));

var _switch_synced = _interopRequireDefault(require("./switch_synced"));

var _viewport_constraint = _interopRequireDefault(require("./viewport_constraint"));

var _initialize_block = require("./initialize_block");

var _with_hooks = _interopRequireDefault(require("./with_hooks"));

var _use_loadable = _interopRequireDefault(require("./use_loadable"));

var _use_records = require("./use_records");

var _use_base = _interopRequireDefault(require("./use_base"));

var _use_cursor = _interopRequireDefault(require("./use_cursor"));

var _use_session = _interopRequireDefault(require("./use_session"));

var _use_settings_button = _interopRequireDefault(require("./use_settings_button"));

var _use_synced = _interopRequireDefault(require("./use_synced"));

var _use_watchable = _interopRequireDefault(require("./use_watchable"));

var _use_viewport = _interopRequireDefault(require("./use_viewport"));

var _use_global_config = _interopRequireDefault(require("./use_global_config"));

var _use_view_metadata = _interopRequireDefault(require("./use_view_metadata"));

var _use_record_action_data = _interopRequireDefault(require("./use_record_action_data"));

var _perform_record_action = require("../perform_record_action");

var _unstable_standalone_ui = require("./unstable_standalone_ui");

Object.keys(_unstable_standalone_ui).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _unstable_standalone_ui[key];
    }
  });
});

/** @ignore */
// Import the top-level entry point so that an Sdk instance is created from the
// globally-available AirtableInterface instance. This is a requirement of
// third-party App code. Consumers who cannot provide an implementation of
// AirtableInterface may access a subset of the UI Components available here by
// instead importing `unstable_standalone_ui`.
var globalAlert = new _global_alert.default();
exports.globalAlert = globalAlert;
